intake/catalog/base.py

```diff
@@ -188,11 +206,86 @@ def _close(self):
         # TODO: maybe close all entries?
         pass
 
+class Entries(dict):
+    """Fetches entries from server on item lookup and iteration.
+
+    This fetches pages of entries from the server during iteration and
+    caches them. On __getitem__ it fetches the sepcific entry from the
+    server.
+    """
+    # This has PY3-style lazy methods (keys, values, items). Since it's
+    # internal we should not need the PY2-only iter* variants.
+    def __init__(self, catalog):
+        self._catalog = catalog
+        self._page_cache = collections.OrderedDict()
+        # Put lookups that were due to __getitem__ in a separate cache
+        # so that iteration reflects the server's order, not an
+        # arbitrary cache order.
+        self._direct_lookup_cache = {}
+        self._page_offset = 0
+
+    def reset(self):
+        "Clear caches to force a reload."
+        self._page_cache.clear()
+        self._direct_lookup_cache.clear()
+        self._page_offset = 0
+
+    def __iter__(self):
+        for key in self.keys():
+            yield key
+
+    def __contains__(self, key):
+        # Avoid iterating through all entries.
+        try:
+            self[key]
+        except KeyError:
+            return False
+        else:
+            return True
+
+    def items(self):
+        for item in six.iteritems(self._page_cache):
+            yield item
+        if self._catalog.page_size is None:
+            # We are not paginating, either because the user set page_size=None
+            # or the server is a version of intake before pagination parameters
+            # were added.
+            return
+        # Fetch more entries from the server.
+        while True:
+            page = self._catalog.fetch_page(self._page_offset)
+            self._page_cache.update(page)
```

Comment:
Do we ever want to purge from the head of the cache?

Reply:
Good question, @martindurant. Let me think this through. The purpose of `_page_cache` is to expedite subsequent calls to `items()`. Since those _always_ start from the top -- there is no API to start in the middle -- it makes sense to hold on to the early pages no matter what.

```
iterator1 = cat.items()
# Iterate through some items.
iterator2 = cat.items()
# If have dumped the first page, we'll have to pull it again.
```

We could impose a limit on the number of pages and/entries to cache, as in a new parameter `page_cache_size` alongside `page_size` to make it possible to iterate through unbounded catalogs. I think that's a good idea. But I don't think purging from the head is ever optimal, which is fortunate from the point of view of simplicity.

Does that seem like a good approach to you?

Tangentially related:
* I have a `search` branch based off of this branch that supports search on remote catalogs (and nested catalogs therein) which will also be useful for navigating through large catalogs, addressing that concern about "there is no API to start in the middle" mentioned above. 
* As you likely already noticed, the _entire_ cache is occasionally purged, of course, whenever the Catalog reloads.

Reply:
Happy to leave it. Perhaps at some point someone will want a purely iterative approach, or partitioned entry list to process with dask... but no need for now.